fineract-provider/src/test/java/org/apache/fineract/infrastructure/core/TenantMigrationFlywayFailNotLatestVersionScenario.java

```diff
@@ -0,0 +1,103 @@
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.infrastructure.core;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.BDDMockito.verify;
+import static org.mockito.BDDMockito.verifyNoInteractions;
+import static org.mockito.BDDMockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.mock;
+
+import com.zaxxer.hikari.HikariDataSource;
+import java.util.List;
+import javax.sql.DataSource;
+import org.apache.fineract.infrastructure.core.config.FineractProperties;
+import org.apache.fineract.infrastructure.core.domain.FineractPlatformTenant;
+import org.apache.fineract.infrastructure.core.service.migration.ExtendedSpringLiquibase;
+import org.apache.fineract.infrastructure.core.service.migration.ExtendedSpringLiquibaseFactory;
+import org.apache.fineract.infrastructure.core.service.migration.SchemaUpgradeNeededException;
+import org.apache.fineract.infrastructure.core.service.migration.TenantDataSourceFactory;
+import org.apache.fineract.infrastructure.core.service.migration.TenantDatabaseStateVerifier;
+import org.apache.fineract.infrastructure.core.service.migration.TenantDatabaseUpgradeService;
+import org.apache.fineract.infrastructure.security.service.TenantDetailsService;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class TenantMigrationFlywayFailNotLatestVersionScenario {
+
+    @Autowired
+    private TenantDetailsService tenantDetailsService;
+
+    @Autowired
+    private HikariDataSource tenantDataSource;
+
+    @Autowired
+    private FineractProperties fineractProperties;
+
+    @Autowired
+    private TenantDatabaseStateVerifier databaseStateVerifier;
+
+    @Autowired
+    private ExtendedSpringLiquibaseFactory liquibaseFactory;
+
+    @Autowired
+    private TenantDataSourceFactory tenantDataSourceFactory;
+
+    @Autowired
+    private TenantDatabaseUpgradeService tenantDatabaseUpgradeService;
+
+    public TenantMigrationFlywayFailNotLatestVersionScenario() throws Exception {
+        // TODO: @galovics need your help here
+
+        // given
+        ExtendedSpringLiquibase initialTenantStoreLiquibase = mock(ExtendedSpringLiquibase.class);
```

Comment:
```
Given("Liquibase is enabled with a default tenant", () -> {
initialTenantStoreLiquibase = mock(ExtendedSpringLiquibase.class);
tenantStoreLiquibase = mock(ExtendedSpringLiquibase.class);

initialTenantLiquibase = mock(ExtendedSpringLiquibase.class);
tenantLiquibase = mock(ExtendedSpringLiquibase.class);

defaultTenant = mock(FineractPlatformTenant.class);
allTenants = List.of(defaultTenant);
defaultTenantDataSource = mock(DataSource.class);

given(databaseStateVerifier.isLiquibaseDisabled()).willReturn(false);
given(fineractProperties.getTenant()).willReturn(new FineractProperties.FineractTenantProperties());
        given(liquibaseFactory.create(tenantDataSource, "tenant_store_db", "initial_switch")).willReturn(initialTenantStoreLiquibase);
        given(liquibaseFactory.create(tenantDataSource, "tenant_store_db")).willReturn(tenantStoreLiquibase);

given(tenantDetailsService.findAllTenants()).willReturn(allTenants);
       given(tenantDataSourceFactory.create(defaultTenant)).willReturn(defaultTenantDataSource);
        given(liquibaseFactory.create(defaultTenantDataSource, "tenant_db", "initial_switch")).willReturn(initialTenantLiquibase);
        given(liquibaseFactory.create(defaultTenantDataSource, "tenant_db")).willReturn(tenantLiquibase);
}
Given("Liquibase runs the very first time for the tenant store", () -> {
given(databaseStateVerifier.isFirstLiquibaseMigration(tenantDataSource)).willReturn(true);
}
Given("A previously Flyway migrated tenant store database", () -> {
given(databaseStateVerifier.isFlywayPresent(tenantDataSource)).willReturn(true);
}
Given("A previously Flyway migrated tenant store database on an earlier version than 1.6", () -> {
given(databaseStateVerifier.isTenantStoreOnLatestUpgradableVersion(tenantDataSource)).willReturn(false);
}

Given("Liquibase runs the very first time for the default tenant", () -> {
given(databaseStateVerifier.isFirstLiquibaseMigration(defaultTenantDataSource)).willReturn(true);
}
Given("A previously Flyway migrated default tenant database", () -> {
given(databaseStateVerifier.isFlywayPresent(defaultTenantDataSource)).willReturn(true);
}
Given("A previously Flyway migrated default tenant database on the 1.6 version", () -> {
given(databaseStateVerifier.isTenantStoreOnLatestUpgradableVersion(defaultTenantDataSource)).willReturn(true);
}
```
And we can change the `When("The database migration process is executed")` with the capability to handle exceptions so later in a Then we can verify it.

```
Then("The tenant store upgrade fails with a schema upgrade needed", () -> {
assertThat(result).isNotNull();
        verify(liquibaseFactory).create(eq(tenantDataSource), any());
        verifyNoMoreInteractions(liquibaseFactory);
        verifyNoInteractions(initialTenantStoreLiquibase, tenantStoreLiquibase, initialTenantLiquibase, tenantLiquibase);
}
```

Sorry for the formatting, but you get the idea.

Reply:
@galovics  A detail or two might need some ironing... but more or less I have it... we are now effectively testing Liquibase enabled/disabled with parameters (and one step definition class). I've left 1-2 TODOs, if you could have a look at "LiquibaseStepDefinitions" that would be great to finish this off.